File: 0016.3Sum_Closest/solution.py
class Solution:
    def three_sum_closest(self, nums, target):
        nums = sorted(nums)
        ans = sum(nums[:3])

        for i in range(len(nums) - 2):
            j, k = i + 1, len(nums) - 1

            while j < k:
                s = nums[i] + nums[j] + nums[k]

                if abs(ans - target) > abs(s - target):
                    ans = s

                if s < target:
                    j += 1
                elif s > target:
                    k -= 1
                else:
                    return s

        return ans


if __name__ == "__main__":
    nums = [1, 0, -1, 3, -2, 1]
    assert Solution().three_sum_closest(nums, 2) == 2

File: 0016.3Sum_Closest/solution.go
package leetcode0016

import "sort"

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	length := len(nums)
	if length < 3 {
		return 0
	}
	ans := nums[0] + nums[1] + nums[2]
	if length == 3 {
		return ans
	}
	for i := 0; i < length-2; i++ {
		l, r := i+1, length-1
		for l < r {
			s := nums[i] + nums[l] + nums[r]
			if abs(s-target) < abs(ans-target) {
				ans = s
			}
			if s < target {
				l++
			} else if s > target {
				r--
			} else {
				return target
			}
		}
	}
	return ans
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
